File: ucharm/compat/fnmatch.py
# ucharm/compat/fnmatch.py
"""
Pure Python implementation of fnmatch for MicroPython.

Provides Unix filename pattern matching WITHOUT using regex,
since MicroPython's regex engine is limited.

Pattern syntax:
- * matches everything
- ? matches any single character
- [seq] matches any character in seq
- [!seq] matches any character not in seq

Note: Patterns are case-sensitive by default.
"""


def _match_pattern(name, pattern):
    """
    Match name against pattern using recursive approach.

    Returns True if name matches pattern.
    """
    ni = 0  # name index
    pi = 0  # pattern index
    nlen = len(name)
    plen = len(pattern)

    # For backtracking on * matches
    star_pi = -1
    star_ni = -1

    while ni < nlen:
        if pi < plen:
            pc = pattern[pi]

            if pc == "?":
                # ? matches any single character
                ni += 1
                pi += 1
                continue

            elif pc == "*":
                # * matches zero or more characters
                # Remember this position for backtracking
                star_pi = pi
                star_ni = ni
                pi += 1
                continue

            elif pc == "[":
                # Character class
                pi += 1
                if pi >= plen:
                    return False

                # Check for negation
                negate = False
                if pattern[pi] == "!":
                    negate = True
                    pi += 1
                elif pattern[pi] == "^":
                    negate = True
                    pi += 1

                # Find closing bracket and check if char matches
                matched = False
                start_pi = pi

                while pi < plen and pattern[pi] != "]":
                    # Check for range like a-z
                    if (
                        pi + 2 < plen
                        and pattern[pi + 1] == "-"
                        and pattern[pi + 2] != "]"
                    ):
                        lo = pattern[pi]
                        hi = pattern[pi + 2]
                        if lo <= name[ni] <= hi:
                            matched = True
                        pi += 3
                    else:
                        if pattern[pi] == name[ni]:
                            matched = True
                        pi += 1

                if pi >= plen:
                    # No closing bracket found
                    return False

                pi += 1  # Skip closing bracket

                if negate:
                    matched = not matched

                if not matched:
                    # No match, try backtracking
                    if star_pi >= 0:
                        pi = star_pi + 1
                        star_ni += 1
                        ni = star_ni
                        continue
                    return False

                ni += 1
                continue

            elif pc == name[ni]:
                # Exact character match
                ni += 1
                pi += 1
                continue

        # No match at current position
        # Try backtracking if we had a * earlier
        if star_pi >= 0:
            pi = star_pi + 1
            star_ni += 1
            ni = star_ni
            continue

        return False

    # Consume any trailing *
    while pi < plen and pattern[pi] == "*":
        pi += 1

    return pi == plen


def fnmatch(name, pattern):
    """
    Test whether name matches pattern.

    Patterns are Unix shell style:
    - * matches everything
    - ? matches any single character
    - [seq] matches any character in seq
    - [!seq] matches any character not in seq

    Example:
        fnmatch('foo.txt', '*.txt')  # True
        fnmatch('bar.py', '*.txt')   # False
    """
    return _match_pattern(name, pattern)


def fnmatchcase(name, pattern):
    """
    Test whether name matches pattern, case-sensitively.

    This is the case-sensitive version of fnmatch.
    """
    return _match_pattern(name, pattern)


def filter(names, pattern):
    """
    Return the subset of names that match pattern.

    Example:
        filter(['a.txt', 'b.py', 'c.txt'], '*.txt')  # ['a.txt', 'c.txt']
    """
    result = []
    for name in names:
        if _match_pattern(name, pattern):
            result.append(name)
    return result


def translate(pattern):
    """
    Translate a shell pattern to a regular expression.

    Note: This is provided for API compatibility but the regex
    may not work on MicroPython's limited regex engine.
    """
    res = []
    i = 0
    n = len(pattern)

    while i < n:
        c = pattern[i]
        i += 1

        if c == "*":
            res.append(".*")
        elif c == "?":
            res.append(".")
        elif c == "[":
            j = i
            if j < n and pattern[j] == "!":
                j += 1
            if j < n and pattern[j] == "]":
                j += 1
            while j < n and pattern[j] != "]":
                j += 1
            if j >= n:
                res.append("\\[")
            else:
                stuff = pattern[i:j]
                if stuff[0:1] == "!":
                    stuff = "^" + stuff[1:]
                res.append("[" + stuff + "]")
                i = j + 1
        else:
            # Escape special characters
            if c in ".^$+{}[]|()\\?":
                res.append("\\")
            res.append(c)

    return "(?s:" + "".join(res) + ")\\Z"

File: native/fnmatch/fnmatch.zig
// fnmatch.zig - Native fnmatch module for ucharm
//
// This module provides Unix shell-style pattern matching:
// - fnmatch(name, pattern) - Test if name matches pattern
// - fnmatchcase(name, pattern) - Case-sensitive matching
// - filter(names, pattern) - Filter list by pattern
// - translate(pattern) - Convert pattern to regex

const std = @import("std");

// ============================================================================
// Pattern Matching Implementation
// ============================================================================

/// Match a name against a shell-style pattern
/// Supports: * (match everything), ? (match one char), [seq] (char class), [!seq] (negated)
pub fn fnmatch_impl(name: [*:0]const u8, pattern: [*:0]const u8, case_sensitive: bool) bool {
    var ni: usize = 0;
    var pi: usize = 0;

    while (pattern[pi] != 0) {
        const pc = pattern[pi];

        if (pc == '*') {
            // Skip consecutive stars
            while (pattern[pi] == '*') {
                pi += 1;
            }

            // Trailing * matches everything
            if (pattern[pi] == 0) {
                return true;
            }

            // Try to match rest of pattern at each position
            while (name[ni] != 0) {
                if (fnmatch_impl(name + ni, pattern + pi, case_sensitive)) {
                    return true;
                }
                ni += 1;
            }
            return fnmatch_impl(name + ni, pattern + pi, case_sensitive);
        } else if (pc == '?') {
            // Match any single character
            if (name[ni] == 0) {
                return false;
            }
            ni += 1;
            pi += 1;
        } else if (pc == '[') {
            // Character class - first check if there's a closing bracket
            var has_closing = false;
            var check_pi = pi + 1;
            while (pattern[check_pi] != 0) {
                if (pattern[check_pi] == ']') {
                    has_closing = true;
                    break;
                }
                check_pi += 1;
            }

            // If no closing bracket, treat '[' as literal
            if (!has_closing) {
                if (name[ni] == 0 or name[ni] != '[') {
                    return false;
                }
                ni += 1;
                pi += 1;
                continue;
            }

            if (name[ni] == 0) {
                return false;
            }

            pi += 1;
            const negated = pattern[pi] == '!' or pattern[pi] == '^';
            if (negated) {
                pi += 1;
            }

            var matched = false;
            const nc = if (case_sensitive) name[ni] else to_lower(name[ni]);

            while (pattern[pi] != 0 and pattern[pi] != ']') {
                const range_start = if (case_sensitive) pattern[pi] else to_lower(pattern[pi]);
                pi += 1;

                if (pattern[pi] == '-' and pattern[pi + 1] != ']' and pattern[pi + 1] != 0) {
                    pi += 1;
                    const range_end = if (case_sensitive) pattern[pi] else to_lower(pattern[pi]);
                    pi += 1;

                    if (nc >= range_start and nc <= range_end) {
                        matched = true;
                    }
                } else {
                    if (nc == range_start) {
                        matched = true;
                    }
                }
            }

            if (pattern[pi] == ']') {
                pi += 1;
            }

            if (matched == negated) {
                return false;
            }
            ni += 1;
        } else {
            // Literal character match
            if (name[ni] == 0) {
                return false;
            }

            const nc = if (case_sensitive) name[ni] else to_lower(name[ni]);
            const expected = if (case_sensitive) pc else to_lower(pc);

            if (nc != expected) {
                return false;
            }
            ni += 1;
            pi += 1;
        }
    }

    return name[ni] == 0;
}

fn to_lower(c: u8) u8 {
    if (c >= 'A' and c <= 'Z') {
        return c + 32;
    }
    return c;
}

// ============================================================================
// Translate Pattern to Regex
// ============================================================================

/// Translate shell pattern to regex pattern
/// Writes to provided buffer, returns length written
pub fn translate_impl(pattern: [*:0]const u8, buf: [*]u8, buf_size: usize) usize {
    var pi: usize = 0;
    var bi: usize = 0;

    // Start with anchor
    if (bi < buf_size) {
        buf[bi] = '(';
        bi += 1;
    }
    if (bi < buf_size) {
        buf[bi] = '?';
        bi += 1;
    }
    if (bi < buf_size) {
        buf[bi] = 's';
        bi += 1;
    }
    if (bi < buf_size) {
        buf[bi] = ')';
        bi += 1;
    }

    while (pattern[pi] != 0 and bi + 5 < buf_size) {
        const c = pattern[pi];

        if (c == '*') {
            buf[bi] = '.';
            bi += 1;
            buf[bi] = '*';
            bi += 1;
        } else if (c == '?') {
            buf[bi] = '.';
            bi += 1;
        } else if (c == '[') {
            buf[bi] = '[';
            bi += 1;
            pi += 1;

            if (pattern[pi] == '!') {
                buf[bi] = '^';
                bi += 1;
                pi += 1;
            } else if (pattern[pi] == '^') {
                buf[bi] = '\\';
                bi += 1;
                buf[bi] = '^';
                bi += 1;
                pi += 1;
            }

            while (pattern[pi] != 0 and pattern[pi] != ']' and bi + 2 < buf_size) {
                buf[bi] = pattern[pi];
                bi += 1;
                pi += 1;
            }

            if (pattern[pi] == ']') {
                buf[bi] = ']';
                bi += 1;
            }
        } else if (c == '.' or c == '+' or c == '^' or c == '$' or
            c == '(' or c == ')' or c == '{' or c == '}' or
            c == '|' or c == '\\')
        {
            // Escape regex special characters
            buf[bi] = '\\';
            bi += 1;
            buf[bi] = c;
            bi += 1;
        } else {
            buf[bi] = c;
            bi += 1;
        }

        pi += 1;
    }

    // End anchor
    if (bi < buf_size) {
        buf[bi] = '\\';
        bi += 1;
    }
    if (bi < buf_size) {
        buf[bi] = 'Z';
        bi += 1;
    }

    if (bi < buf_size) {
        buf[bi] = 0;
    }

    return bi;
}

// ============================================================================
// Exported Functions (C ABI)
// ============================================================================

/// Match name against pattern (case-sensitive on Unix, like CPython)
pub export fn fnmatch_fnmatch(name: [*:0]const u8, pattern: [*:0]const u8) bool {
    return fnmatch_impl(name, pattern, true);
}

/// Match name against pattern (case-sensitive)
pub export fn fnmatch_fnmatchcase(name: [*:0]const u8, pattern: [*:0]const u8) bool {
    return fnmatch_impl(name, pattern, true);
}

/// Translate pattern to regex, write to buffer
pub export fn fnmatch_translate(pattern: [*:0]const u8, buf: [*]u8, buf_size: usize) usize {
    return translate_impl(pattern, buf, buf_size);
}

// ============================================================================
// Tests
// ============================================================================

test "fnmatch basic" {
    try std.testing.expect(fnmatch_fnmatch("hello.txt", "*.txt"));
    try std.testing.expect(fnmatch_fnmatch("hello.txt", "hello.*"));
    try std.testing.expect(fnmatch_fnmatch("hello.txt", "h*o.txt"));
    try std.testing.expect(!fnmatch_fnmatch("hello.txt", "*.py"));
}

test "fnmatch question mark" {
    try std.testing.expect(fnmatch_fnmatch("hello", "hell?"));
    try std.testing.expect(fnmatch_fnmatch("hello", "h?llo"));
    try std.testing.expect(!fnmatch_fnmatch("hello", "h?lo"));
}

test "fnmatch character class" {
    try std.testing.expect(fnmatch_fnmatch("hello", "[gh]ello"));
    try std.testing.expect(fnmatch_fnmatch("hello", "[a-z]ello"));
    // Case-sensitive fnmatch: [A-Z] does NOT match lowercase
    try std.testing.expect(!fnmatch_fnmatch("hello", "[A-Z]ello"));
    try std.testing.expect(!fnmatch_fnmatchcase("hello", "[A-Z]ello"));
}

test "fnmatch negated class" {
    try std.testing.expect(fnmatch_fnmatch("hello", "[!abc]ello"));
    try std.testing.expect(!fnmatch_fnmatch("hello", "[!h]ello"));
}

test "fnmatchcase" {
    try std.testing.expect(fnmatch_fnmatchcase("Hello", "Hello"));
    try std.testing.expect(!fnmatch_fnmatchcase("Hello", "hello"));
    try std.testing.expect(!fnmatch_fnmatch("Hello", "hello")); // case-sensitive on Unix
}
